skimage/feature/scattering.py

```diff
@@ -0,0 +1,286 @@
+    phi : Low pass filter as a 2D ndarray
+        Low pass filter in the Fourier domain, stored as (n,n) matrix
+    J : int
+        Rate of subsampling 2**J
+    n_scat : int
+        number of spatial coefficients of the scattering vector
+    Returns
+    -------
+    XX : ndarray
+        stacked images after being filtered and subsampled
+    """
+    img_filtered = np.real(np.fft.ifft2(np.fft.fft2(img)*phi))
+    n_nolp = img.shape[-1]
+    ds = int(n_nolp / n_scat)
+    return 2 ** (J - 1) * img_filtered[..., ::ds, ::ds]
+
+
+def scattering(x, wavelet_filters=None, m=2):
+    """Compute the scattering transform of a signal (or set of signals).
+    Given 'x', a set of 2D signals, this function computes the
+    scattering transform of these signals using the filter bank
+    'wavelet_filters'.
+    Notes
+    -----
+    **Bondary values: The scattering transform applies a set of convolutions
+    to the input signals. These convolutions are computed as the point-wise
+    multiplication in the Fourier domain, thus the boundary values of the
+    image are circular or cyclic. In case you need other
+    kind of boundary values, for instance zero-padded, you should
+    edit the images before calling this function.
+    **Shape of x: The signals x must be squared shaped, thus (N,px,px)
+    and not (N,px,py) for py != px. In case the images are rectangular,
+    they will be cropped to the smallest dimension, px or py.
+    Parameters
+    ----------
+    x : array_like
+        3D dnarray with N images (2D arrays) of size (px,px), thus x has size
+         (N,px,px)
+        In case the array is rectangular (N,px,py) for px not equal to py, t
+        he images will be cropped.
+    wavelet_filters : Dictionary with the multiresolution wavelet filter bank
+        Dictionary of vectors obtained after calling:
+            >>>> px = 32 #number of pixels of the images
+            >>>> J = np.log2(px) #number of scales
+            >>>> wavelet_filters = multiresolution_filter_bank_morlet2d(px,J=J)
+    m : int
+        Order of the scattering transform, which can be 0, 1 or 2.
+    Returns
+    -------
+    S : 4D array_like
+        Scattering transform of the x signals, of size
+        (N,num_coeffs,spatial_coefs,spatial_coefs). For more information
+        see _[1] _[2]
+    U : array_like
+        Result before applying the lowpass filter and subsampling.
+    S_tree : dictionary
+        Dictionary that allows to access the scattering coefficients (S)
+        according to the layer and indices. More specifically:
+            Zero-order layer: The only available key is 0:
+            S_tree[0] : returns all the coefficients of the 0-order scattering
+             transform
+
+            First-order layer: The keys are tuples with (j,l) indexing
+            S_tree[(j,l)] : returns the first order coefficients for scale 'j'
+             and angle 'l'
+            S_tree[((j1,l1),(j2,l2))] : the second order for scale 'j1',
```

Comment:
There are two things: What you have written `S_tree[((j1, l1), (j2, l2))]` is equivalent to `S_tree[(j1, l1), (j2, l2)]`. And that way you would semantically group the parameters pertaining to each layer.

I prefer flattening the indices to make the thing feel like an array, because it essentially is half of one.